Rename forecast outputs

isaac-tfn opened the pull request:
Rename outputs from ntem forecasting to include scenario in file names

fixes #985

mishtaiwi committed:
Rewrite EDGE replicant in numpy

mishtaiwi committed:
Sectorise Matrices Script

mishtaiwi committed:
Re: #974

- Rename headers of output CSVs
- Sort CSVs to avoid csv>mat errors

mishtaiwi committed:
Check for not grown demand proportion.

isaac-tfn committed:
Small changes for extractor to work

isaac-tfn committed:
Created module for preprocessing growth_matrices
Created module for handling ticket splits
Created module for general utils (mainly matrix manipulations and conversions)
Deleted functions from edge_replicant as they're moved to new modules
Nothing new tested yet

isaac-tfn committed:
Run through to get processed and infilled growth matrices and ticket type splits.
Need to look at caching preprocessing steps.

isaac-tfn committed:
minor change to config yml used for testing

isaac-tfn committed:
Multiprocessing is working - code needs tidying up.

isaac-tfn committed:
Changed multiprocessing so that it doesn't need to write out intermediary variables and read them back in. Small change to config class to make sure export dir exists.

isaac-tfn committed:
Couple of changes, moving to vm

isaac-tfn committed:
Tidied up a bit and added some documentation

isaac-tfn committed:
Mid changes on forecast file names

Committing to change branch, currently initial PA matrices are renamed with scenario but that hasn't carried through to other matrices

isaac-tfn committed:
Created module for loading in global variables and a class for storing them. Altered ticket splits to be able to load directly if available.

isaac-tfn committed:
Ticket type splits are cached when created, and the config can be altered to load this cache in instead of generating splits internally, saving a few minutes

isaac-tfn committed:
Altered run_forecast to run new process instead of edge_replicant for edge. Changed import structure of EDGE_growth.

isaac-tfn committed:
Documenting for a config class

isaac-tfn committed:
Fixed typo in class name

isaac-tfn committed:
Deleted part of run script. Runs from run_forecast now

isaac-tfn committed:
Outputs now named and forecasting runs

isaac-tfn committed:
Crude check for existing ticket splits if not provided explicitly

isaac-tfn committed:
Added documentation to config module

isaac-tfn committed:
fji

isaac-tfn committed:
changes for testing

BenTaylor-TfN committed:
Re-ran Black with correct config

BenTaylor-TfN commented on normits_demand/utils/general.py:
```diff
@@ -1578,27 +1586,26 @@ def ensure_segmentation(df: pd.DataFrame,
+    m_list: Iterable[int],
+    ca_list: Iterable[int],
+    tp_list: Iterable[int] = None,
+) -> (Union[Iterator[Tuple[str, str, int, int, int]], Iterator[Tuple[str, str, int, int]],]):
```
Remove the comma at the end of this line

BenTaylor-TfN commented on normits_demand/utils/general.py:
```diff
@@ -2157,25 +2150,24 @@ def is_in_string(vals: Iterable[str],
-                             compress: bool = False,
-                             suffix: str = None,
-                             ) -> str:
+def get_compiled_matrix_name(
```
How is this function different to `get_dist_name`? 

I realise you might not be the best person to answer this question - but an insight into what you've discovered could be useful here.
I imagine they're similar functions used by different parts of the code. Is there any obvious reason why?

isaac-tfn replied:
It seems likely thst whoever wrote this (I think Matt) possibly forgot they'd already made such a similar function, and at a minimum this function should inherit from get_dist_name. Possibly out of scope for the current PR to tidy up.

BenTaylor-TfN commented on normits_demand/models/forecasting/ntem_forecast.py:
```diff
@@ -172,6 +175,7 @@ def output_filename(
         trip_origin: str,
         purpose: int,
         year: int,
+        scenario: str = "Core",
```
Does it makes sense to default to "Core" here? Would it make more sense to force the user to pick a name?

I'm concerned that with all the new scenarios in NTEM data, this could lead to unexpected errors.

This addition is also not documented

isaac-tfn replied:
Agree don't know why I did this.

BenTaylor-TfN commented on normits_demand/models/forecasting/ntem_forecast.py:
```diff
@@ -642,6 +656,7 @@ def grow_all_matrices(
     matrices: NTEMImportMatrices,
     growth: TEMProTripEnds,
     output_folder: Path,
+    scenario: str = "TEM",
```
Does this need to have a default value? I'm not convinced this makes obvious sense.
This is probably linked to the same question I have in `run_forecast`

isaac-tfn replied:
Yes should probably just be none for TEM forecasting.

BenTaylor-TfN commented on normits_demand/models/forecasting/ntem_forecast.py:
```diff
@@ -807,7 +831,10 @@ def compile_highway_for_rail(pa_folder: Path, years: list[int], mode: list[int])
+    import_od_path: Path,
+    years: List[int],
+    car_occupancies_path: Path,
+    scenario: str = None,
```
Undocumented argument

BenTaylor-TfN reviewed (commented):
Overall notes:
- When using black on code in demand, I would advise doing this in a separate PR, or just turning it off when working in Demand. Most of the code doesn't meet that standard. Otherwise:
    - This makes it really hard to see what you've really done
    - You're now in the blame as the last person to edit those lines of code if something goes wrong!
- I think your implementation of black isn't picking up the proper demand config - it looks like your implementation is being too aggressive on the line lengths, making the code more difficult to read - I've re-ran black to make the code easier to review.
- I haven't reviewed `ntem_forecast_checks` - perhaps @wsp-mbuckley  is best suited to that once my above comments have been dealt with

isaac-tfn commented on normits_demand/core/segments.py:
```diff
@@ -326,7 +333,10 @@ def _mul_div_segmentation(self,
                 "multiplication.\n"
                 "Expected %s segments.\n"
                 "Found %s segments."
-                % (len(return_seg.segment_names), len(set(multiply_dict.keys())))
+                % (
+                    len(return_seg.segment_names),
+                    len(set(multiply_dict.keys())),
+                )
```
I have always assumed there must be a good reason to format like this instead of f strings so I tend to not change when I see them, but will change now!

isaac-tfn commented on normits_demand/core/segments.py:
```diff
@@ -1094,15 +1128,20 @@ def aggregate(self,
-                zip(self.segment_names, [[n] for n in self.segment_names])
+                zip(
+                    self.segment_names,
+                    [[n] for n in self.segment_names],
```
I think this is something black does itself in order to split across multiple lines - it's just my overaggressive black settings which I've now turned off

isaac-tfn commented on normits_demand/core/segments.py:
```diff
@@ -2019,14 +2051,17 @@ def generate_template_file_name(
         if suffix is not None:
             name_parts += [suffix]
 
+        if scenario is not None:
+            name_parts += [scenario]
+
```
I think it makes more sense to have concrete specified filename parts where they exist and leave suffix as a misc catcher for more specific cases. It is an oversight to not have kept 'suffix' as the end of the file name though!

isaac-tfn commented on normits_demand/utils/general.py:
```diff
@@ -930,6 +928,9 @@ def get_dist_name(trip_origin: str = None,
     if not is_none_like(mode):
         name_parts += [f"m{mode}"]
 
+    if not is_none_like(scenario):
+        name_parts += [f"{scenario}"]
+
```
These two areas don't match up in terms of parts in general - needs more investigating

isaac-tfn commented on run_models/run_forecast.py:
```diff
@@ -225,6 +239,7 @@ def tem_forecasting(
         tripend_data = tem_forecast.read_tripends(
             params.base_year, params.future_years, params.tripend_path
         )
+        scenario = "TEM"
```
As mentioned I think this duplicates a comment elsewhere and yes I agree

isaac-tfn committed:
Small name changes

isaac-tfn committed:
Merge branch 'EDGE_growth_refactor' of https://github.com/Transport-for-the-North/NorMITs-Demand into EDGE_growth_refactor

isaac-tfn committed:
Some changes made from comments. Main discussion on the multiple very similar functions (get_dist_name etc.). Will discuss at a later date whether it's worth the time to consolidate them as I think there's unnecessary duplication but I didn't write any of them so don't want to change anything just yet.

isaac-tfn committed:
Mid change commit to switch to vm

isaac-tfn committed:
Outputs match previous iteration. Not all comments addressed.

isaac-tfn committed:
All messages resolved. Not all dealt with but committing what is to start working through issues.

isaac-tfn committed:
Created new dataclass for return from _tp_loop

isaac-tfn committed:
Added in two new dirs to config:
ticket_splits_dir is for ticket splits and/or files for creating them
irsj_props_dir is for irsj_probabilty files
These can be set to be equal to matrices_to_grow_dir, as currently all of these files are in that directory

isaac-tfn committed:
Removed 'missing_factors' as outputs were messy and there are multiple other checks for this.

isaac-tfn committed:
small change

isaac-tfn committed:
Added technology scenario, and extracted databases (not part of this commit or repo)

isaac-tfn committed:
Removed config_path as argument.

Files have been copied to data path, can be removed soon.

isaac-tfn committed:
Fixed some small changes around multiprocessing outputs

isaac-tfn committed:
Removed unused imports

isaac-tfn committed:
Tiny change to a string format

BenTaylor-TfN committed:
Merge branch 'master' into EDGE_growth_refactor

BenTaylor-TfN committed:
Review edits

BenTaylor-TfN committed:
Update norms_stn_forecast.yml

isaac-tfn committed:
Change to netm forecast yml

isaac-tfn committed:
Merge branch 'EDGE_growth_refactor' of https://github.com/Transport-for-the-North/NorMITs-Demand into EDGE_growth_refactor

isaac-tfn committed:
Fixed issue with toolkit imports

isaac-tfn committed:
Getting edge forecasting working

isaac-tfn committed:
Switched to using toolkit translation

isaac-tfn committed:
Added warning for merge fail in tickettypesplits

isaac-tfn committed:
applicator code running

isaac-tfn committed:
Function to apply covid adjustments

isaac-tfn committed:
Add cap year factor

isaac-tfn committed:
Fix out_year in covid script

isaac-tfn committed:
Merge pull request #998 from Transport-for-the-North/EDGE_growth_refactor

Edge growth refactor

isaac-tfn committed:
Merge branch 'forecast_changes' into rename_forecast_outputs
